src/components/utils/src/threads/thread_posix.cc

```diff
@@ -261,15 +270,27 @@ void Thread::stop() {
   SDL_DEBUG("Stopped thread #" << handle_ << " \"" << name_ << " \"");
 }
 
-void Thread::join() {
+void Thread::join(const bool force_stop) {
   SDL_AUTO_TRACE();
   DCHECK_OR_RETURN_VOID(!IsCurrentThread());
 
-  stop();
+  if (force_stop) {
+    stop();
+  }
 
   sync_primitives::AutoLock auto_lock(state_lock_);
+
+  if (!force_stop) {  // then wait on going iteration to finish
+    while (do_iteration_ && thread_created_) {
+      state_cond_.Wait(auto_lock);
+    }
+  }
+
+  stopped_ = true;
+  do_iteration_ = true;
   run_cond_.NotifyOne();
-  if (isThreadRunning_) {
+
+  while (do_iteration_ && thread_created_) {
     state_cond_.Wait(auto_lock);
   }
 }
```

Comment:
@orhan-mehmedov it's not clear for me why are you going to wait `state_cond_` condvar twice? And one more important thing:
I see that there is a possibility to wait for `state_cond_` condvar from different threads simulteniously (f.e. here and in `start()` method). What happens in case condvar will be notified? BTW I see that it could be notified with `NotifyOne()`. Which thread would be notified?